internal/controller/ownerref.go

```diff
@@ -73,3 +74,65 @@ func findFinalOwnerRef(ctx context.Context, client crclient.Client, ns string, o
 	Log.V(1).Info("Pods owner found", "Kind", ownerRef.Kind, "Name", ownerRef.Name, "Namespace", ns)
 	return ownerRef, nil
 }
+
+//FindControllerOwner returns OwnerReferent that owns controller process
+//it starts searching from the current pod and then resolves owners recursively
+//until object without owner is not found
+func FindControllerDeployment(ctx context.Context, client crclient.Client) (*v1.Deployment, error) {
+	ns, err := k8sutil.GetOperatorNamespace()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get current Pod the operator is running in
+	pod, err := k8sutil.GetPod(ctx, client, ns)
+	if err != nil {
+		return nil, err
+	}
+	// Get Owner that the Pod belongs to
+	//TODO Take a look, maybe it's a better thing to use
+	//podOwnerRefs := metav1.NewControllerRef(pod, pod.GroupVersionKind())
+	ownerRef := metav1.GetControllerOf(pod)
+	deployment, err := findDeployment(ctx, client, ns, ownerRef)
```

Comment:
I would move this first ownerref into `findDeployment` -- it doesn't make as much sense to start one step up the ladder instead of at the base:

```go
deployment, err := findDeployment(crx, client, ns, pod)
// ...
func findDeployment(...) {
  ownerRef := metav1.GetControllerOf(pod)
  // continue as normal
}
```
I believe you can use a type such as `metav1.Object` for this purpose and then reflect the final result to a Deployment.

Reply:
(This wasn't my original code, all the changes I took in ownerrefs was from an old closed PR on this repo) but I think the reason why it was done this way is `newOwnerRef` is passed back into findDeployment later so we'd need to pass in the ownerref and not the pod

Reply:
Sure but wouldn't it be cleaner to do

```go
// findDeployment tries to locate the final controller/owner based on the owner reference provided.
func findDeployment(ctx context.Context, client crclient.Client, ns string, metaObj metav1.Object) (*v1.Deployment, error) {
	ownerRef := metav1.GetControllerOf(metaObj)

	if ownerRef == nil {
		return nil, nil
	}

	if ownerRef.Kind == "Deployment" {
		d := &v1.Deployment{}
		err := client.Get(ctx, types.NamespacedName{Namespace: ns, Name: ownerRef.Name}, d)
		d.APIVersion = ownerRef.APIVersion
		d.Kind = ownerRef.Kind
		if err != nil {
			return nil, err
		}
		return d, nil
	}
	obj := &unstructured.Unstructured{}
	obj.SetAPIVersion(ownerRef.APIVersion)
	obj.SetKind(ownerRef.Kind)
	err := client.Get(ctx, types.NamespacedName{Namespace: ns, Name: ownerRef.Name}, obj)
	if err != nil {
		return nil, err
	}

	return findDeployment(ctx, client, ns, obj)
}
```